File: app/src/main/java/org/miaowo/miaowo/fragment/chat/ChatFragment.java
package org.miaowo.miaowo.fragment.chat;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.sdsmdg.tastytoast.TastyToast;

import org.miaowo.miaowo.R;
import org.miaowo.miaowo.adapter.ChatMsgAdapter;
import org.miaowo.miaowo.api.API;
import org.miaowo.miaowo.bean.data.ChatMessage;
import org.miaowo.miaowo.bean.data.ChatMessageList;
import org.miaowo.miaowo.bean.data.ChatRoom;
import org.miaowo.miaowo.custom.ChatListAnimator;
import org.miaowo.miaowo.custom.load_more_list.LoadMoreList;
import org.miaowo.miaowo.root.BaseActivity;
import org.miaowo.miaowo.root.BaseFragment;
import org.miaowo.miaowo.util.HttpUtil;
import org.miaowo.miaowo.util.JsonUtil;
import org.miaowo.miaowo.util.LogUtil;

import java.io.IOException;

import butterknife.BindView;
import butterknife.OnClick;
import okhttp3.FormBody;
import okhttp3.Request;

/**
 * 聊天内容
 * Created by luqin on 17-4-23.
 */

public class ChatFragment extends BaseFragment {
    @BindView(R.id.list) LoadMoreList lv_chat;
    @BindView(R.id.et_msg) EditText et_msg;

    private ChatMsgAdapter mAdapter;
    private ChatRoom mRoom;
    private API mApi;

    public ChatFragment() {}
    public static ChatFragment newInstance(ChatRoom room) {
        ChatFragment fragment = new ChatFragment();
        fragment.mRoom = room;
        return fragment;
    }
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container
            , @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_chat, container, false);
    }
    @Override
    public void initView(View view) {
        HttpUtil mHttp = HttpUtil.utils();
        JsonUtil mJson = JsonUtil.utils();
        mApi = new API();

        mAdapter = new ChatMsgAdapter();
        lv_chat.setAdapter(mAdapter);
        lv_chat.loadMoreControl(false, false);
        lv_chat.setItemAnimator(new ChatListAnimator());

        Request request = new Request.Builder().url(BaseActivity.get.getString(R.string.url_chat_message,
                API.loginUser.getUsername().toLowerCase(), mRoom.getRoomId())).build();
        mHttp.post(request, (call, response) ->
                BaseActivity.get.runOnUiThreadIgnoreError(() -> {
                    try {
                        mAdapter.update(mJson.buildFromAPI(response, ChatMessageList.class).getMessages());
                        lv_chat.scrollToPosition(mAdapter.getItemCount() - 1);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }));
    }

    @OnClick(R.id.btn_send)
    public void onSend() {
        String message = et_msg.getText().toString();
        if (TextUtils.isEmpty(message)) {
            BaseActivity.get.toast(R.string.err_msg_empty, TastyToast.ERROR);
            return;
        }
        FormBody body = new FormBody.Builder()
                .add("message", message)
                .add("timestamp", String.valueOf(System.currentTimeMillis()))
                .add("roomId", String.valueOf(mRoom.getRoomId()))
                .build();
        mApi.useAPI(API.APIType.USERS, mRoom.getLastUser().getUid() + "/chats", API.Method.POST, true, body, (call, response) -> {
            // TODO: API-发送聊天信息
            LogUtil.i(response);
        });
        et_msg.setText("");
    }

    public void newMessage(ChatMessage message) {
        mAdapter.insert(message, false);
        lv_chat.scrollToPosition(mAdapter.getItemCount() - 1);
    }

    public ChatRoom getRoom() {
        return mRoom;
    }
}

File: app/src/main/java/org/miaowo/miaowo/fragment/chat/ChatFragment.kt
package org.miaowo.miaowo.fragment.chat

import android.content.Context
import android.os.Bundle
import android.os.SystemClock
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_chat.*
import org.miaowo.miaowo.R
import org.miaowo.miaowo.adapter.ChatMsgAdapter
import org.miaowo.miaowo.base.extra.handleError
import org.miaowo.miaowo.bean.data.ChatRoom
import org.miaowo.miaowo.other.ChatListAnimator
import org.miaowo.miaowo.other.Const
import org.miaowo.miaowo.API
import org.miaowo.miaowo.Miao
import org.miaowo.miaowo.base.extra.inflateId
import org.miaowo.miaowo.interfaces.IMiaoListener
import java.util.*

/**
 * 聊天内容
 * Created by luqin on 17-4-23.
 */

class ChatFragment : Fragment() {
    private val mAdapter = ChatMsgAdapter()
    private var mRoomId = -1
    private var mUser = -1
    private var mName = ""
    private var mCheckThread: Thread? = null
    private var mListener: IMiaoListener? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is IMiaoListener)
            mListener = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) = inflateId(inflater, R.layout.fragment_chat, container)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        mRoomId = arguments!!.getInt(Const.ID, -1)
        mUser = arguments!!.getInt(Const.USER, -1)
        mName = arguments!!.getString(Const.NAME, "")

        mListener?.setToolbar("聊天: $mName")

        btn_send.setOnClickListener {
            if (et_msg.text.toString().isBlank())
                Miao.i.handleError(R.string.err_no_message)
            else {
                API.Users.chat(mUser, et_msg.text.toString(), mRoomId) {
                    Miao.i.runOnUiThread {
                        if (it != "OK") Miao.i.handleError(it)
                        else list.scrollToPosition(mAdapter.itemCount)
                    }
                }
            }
        }

        list.layoutManager = LinearLayoutManager(context)
        list.adapter = mAdapter
        list.itemAnimator = ChatListAnimator()

        if (mRoomId >= 0) {
            checkMsgUpdate()

            mCheckThread = Thread {
                var isFinish = true
                try {
                    while (true) {
                        while (!isFinish) {
                        }
                        isFinish = false
                        Thread.sleep(5000)
                        checkMsgUpdate {
                            isFinish = true
                        }
                    }
                } catch (e: InterruptedException) {
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            mCheckThread?.start()
        } else fragmentManager?.popBackStackImmediate()
    }

    override fun onDestroy() {
        mCheckThread?.interrupt()
        mCheckThread = null
        super.onDestroy()
    }

    private fun checkMsgUpdate(run: (() -> Unit)? = null) {
        API.Doc.chatMessage(mRoomId) {
            if (isVisible) {
                val receiveList = it?.messages ?: emptyList()
                // 未成功获取数据
                if (receiveList.isEmpty()) return@chatMessage
                // 成功获取数据
                val existLastTime = mAdapter.items.lastOrNull()?.timestamp ?: -1
                val appendList = receiveList.filter { it.timestamp > existLastTime }
                if (appendList.isNotEmpty() && isVisible) mAdapter.append(appendList)
            }
            run?.invoke()
        }
    }

    companion object {
        fun newInstance(room: ChatRoom): ChatFragment {
            val fragment = ChatFragment()
            val args = Bundle()
            args.putInt(Const.ID, room.roomId)
            args.putInt(Const.USER, room.lastUser?.uid ?: -1)
            args.putString(Const.NAME, room.lastUser?.username)
            args.putString(Const.TAG, "${fragment.javaClass.name}.room.${room.roomId}")
            fragment.arguments = args
            return fragment
        }
    }
}
